Bug in Inspector tool. Denominator int

With release 26.4.0 a breaking bug was introduced to the Inspector tool in the calculation of the mean spectrum:


```python
da = <scipp.DataArray>
  Dimensions: Sizes[t:10, y:6, x:6, ]
  Coordinates:
  * t                           int64        <no unit... ..., 1, 1]
  Masks:
    non_finite                   bool        <no unit>  (t, y, x)  [False, False, ..., False, False]
  
  
  rect = {'x': {'dim': 'x', 'value': <scipp.Variable> (vertex: 5)    float64              [m]  [2.5, 5, ..., 2.5, 2.5]}, 'y': {'dim': 'y', 'value': <scipp.Variable> (vertex: 5)    float64              [m]  [2.5, 2.5, ..., 5, 2.5]}}
  op = 'mean'
  
      def _slice_rectangular_region(da: sc.DataArray, rect: dict, op: str) -> sc.DataArray:
          x = rect['x']
          y = rect['y']
          xmin, xmax = x['value'].min(), x['value'].max()
          ymin, ymax = y['value'].min(), y['value'].max()
          try:
              # If there is a 2D coordinate in the data, we need to slice the other dimension
              # first, as trying to slice a 2D coordinate using label-based indexing raises an
              # error in Scipp. After slicing the other dimension, the 2D coordinate will be
              # 1D and can be sliced normally using label-based indexing.
              # We assume here that there would only be one multi-dimensional coordinate in a
              # given DataArray (which is very likely the case).
              if da.coords[y['dim']].ndim > 1:
                  out = da[x['dim'], xmin:xmax][y['dim'], ymin:ymax]
              else:
                  out = da[y['dim'], ymin:ymax][x['dim'], xmin:xmax]
              # If the operation is a mean, there is currently a bug in the implementation
              # in scipp where doing a mean over a subset of the array's dimensions gives the
              # wrong result: https://github.com/scipp/scipp/issues/3841
              # Instead, we manually compute the mean
              dims = (x['dim'], y['dim'])
              if 'mean' not in op:
                  return getattr(out, op)(dims)
              if 'nan' in op:
                  numerator = out.nansum(dims)
                  denominator = (~sc.isnan(out.data)).sum()
              else:
                  numerator = out.sum(dims)
                  denominator = out.size
  >           denominator.unit = ""
              ^^^^^^^^^^^^^^^^
  E           AttributeError: 'int' object has no attribute 'unit' and no __dict__ for setting new attributes
```
@nvaytet (I can't assign you to this bug, apparently?)

--- a/src/plopp/plotting/_inspector.py
+++ b/src/plopp/plotting/_inspector.py
@@ -87,10 +87,10 @@ def _slice_rectangular_region(da: sc.DataArray, rect: dict, op: str) -> sc.DataA
         if 'nan' in op:
             numerator = out.nansum(dims)
             denominator = (~sc.isnan(out.data)).sum()
+            denominator.unit = ""
         else:
             numerator = out.sum(dims)
             denominator = out.size
-        denominator.unit = ""
         return numerator / denominator
     except IndexError:
         # If the index is out of bounds, return an empty DataArray

--- a/src/plopp/plotting/_slicer.py
+++ b/src/plopp/plotting/_slicer.py
@@ -42,6 +42,7 @@ def _maybe_reduce_dim(da, dims, op):
     if 'nan' in op:
         numerator = da.nansum(to_be_reduced)
         denominator = (~sc.isnan(da)).to(dtype=int).sum(to_be_reduced)
+        denominator.unit = ""
     else:
         numerator = da.sum(to_be_reduced)
         denominator = np.prod([da.sizes[dim] for dim in to_be_reduced])

--- a/tests/plotting/inspector_test.py
+++ b/tests/plotting/inspector_test.py
@@ -147,9 +147,8 @@ def test_point_mode_line_data_outside_range():
     assert all(np.isnan(line._data.values))
 
 
-@pytest.mark.usefixtures('_use_ipympl')
-def test_polygon_mode_triangle():
-    da = sc.DataArray(
+def _make_test_data():
+    return sc.DataArray(
         data=sc.array(
             dims=["xx", "yy", "zz"],
             values=[
@@ -167,7 +166,12 @@ def test_polygon_mode_triangle():
             "yy": sc.array(dims=['yy'], values=[0.0, 10.0, 20.0, 30.0], unit='m'),
             "zz": sc.array(dims=['zz'], values=[0.0, 1.0, 2.0], unit='m'),
         },
-    )
+    ).transpose(['zz', 'yy', 'xx'])
+
+
+@pytest.mark.usefixtures('_use_ipympl')
+def test_polygon_mode_triangle():
+    da = _make_test_data()
 
     ip = pp.inspector(da, mode='polygon', dim='zz')
     fig2d = ip[0][0]
@@ -177,8 +181,8 @@ def test_polygon_mode_triangle():
 
     # This triangle should select the bottom left corner of the data.
     # Closing the polygon by repeating the first point.
-    x = [-1, 32, -1, -1]
-    y = [-1, -1, 350, -1]
+    x = [-1, -1, 350, -1]
+    y = [-1, 32, -1, -1]
     for xi, yi in zip(x, y, strict=True):
         tool.click(x=xi, y=yi)
 
@@ -190,7 +194,7 @@ def test_polygon_mode_triangle():
             [False, True, True],
             [True, True, True],
         ],
-    )
+    ).transpose(['yy', 'xx'])
 
     expected = da.assign_masks(m=mask).sum(["xx", "yy"])
     line = next(iter(fig1d.artists.values()))
@@ -199,24 +203,7 @@ def test_polygon_mode_triangle():
 
 @pytest.mark.usefixtures('_use_ipympl')
 def test_polygon_mode_square():
-    da = sc.DataArray(
-        data=sc.array(
-            dims=["xx", "yy", "zz"],
-            values=[
-                [[0, 1, 2], [3, 4, 5], [6, 7, 8]],
-                [[9, 10, 11], [12, 13, 14], [15, 16, 17]],
-                [[18, 19, 20], [21, 22, 23], [24, 25, 26]],
-                [[27, 28, 29], [30, 31, 32], [33, 34, 35]],
-            ],
-        ),
-        coords={
-            "xx": sc.array(
-                dims=['xx'], values=[0.0, 100.0, 200.0, 300.0, 400.0], unit='m'
-            ),
-            "yy": sc.array(dims=['yy'], values=[0.0, 10.0, 20.0, 30.0], unit='m'),
-            "zz": sc.array(dims=['zz'], values=[0.0, 1.0, 2.0], unit='m'),
-        },
-    )
+    da = _make_test_data()
 
     ip = pp.inspector(da, mode='polygon', dim='zz')
     fig2d = ip[0][0]
@@ -225,8 +212,8 @@ def test_polygon_mode_square():
     tool = fig2d.toolbar['inspect']._tool
 
     # This square should select the top right corner of the data.
-    x = [11, 32, 32, 11, 11]
-    y = [101, 101, 410, 410, 101]
+    x = [101, 101, 410, 410, 101]
+    y = [11, 32, 32, 11, 11]
     for xi, yi in zip(x, y, strict=True):
         tool.click(x=xi, y=yi)
 
@@ -238,7 +225,7 @@ def test_polygon_mode_square():
             [True, False, False],
             [True, False, False],
         ],
-    )
+    ).transpose(['yy', 'xx'])
 
     expected = da.assign_masks(m=mask).sum(["xx", "yy"])
     line = next(iter(fig1d.artists.values()))
@@ -247,26 +234,7 @@ def test_polygon_mode_square():
 
 @pytest.mark.usefixtures('_use_ipympl')
 def test_polygon_mode_triangle_with_mask_in_third_dimension():
-    da = sc.DataArray(
-        data=sc.array(
-            dims=["xx", "yy", "zz"],
-            values=[
-                [[0, 1, 2], [3, 4, 5], [6, 7, 8]],
-                [[9, 10, 11], [12, 13, 14], [15, 16, 17]],
-                [[18, 19, 20], [21, 22, 23], [24, 25, 26]],
-                [[27, 28, 29], [30, 31, 32], [33, 34, 35]],
-            ],
-            dtype=float,
-        ),
-        coords={
-            "xx": sc.array(
-                dims=['xx'], values=[0.0, 100.0, 200.0, 300.0, 400.0], unit='m'
-            ),
-            "yy": sc.array(dims=['yy'], values=[0.0, 10.0, 20.0, 30.0], unit='m'),
-            "zz": sc.array(dims=['zz'], values=[0.0, 1.0, 2.0], unit='m'),
-        },
-        masks={'mask': sc.array(dims=['zz'], values=[False, True, False])},
-    )
+    da = _make_test_data()
 
     ip = pp.inspector(da, mode='polygon', dim='zz')
     fig2d = ip[0][0]
@@ -276,8 +244,8 @@ def test_polygon_mode_triangle_with_mask_in_third_dimension():
 
     # This triangle should select the bottom left corner of the data.
     # Closing the polygon by repeating the first point.
-    x = [-1, 32, -1, -1]
-    y = [-1, -1, 350, -1]
+    x = [-1, -1, 350, -1]
+    y = [-1, 32, -1, -1]
     for xi, yi in zip(x, y, strict=True):
         tool.click(x=xi, y=yi)
 
@@ -289,7 +257,7 @@ def test_polygon_mode_triangle_with_mask_in_third_dimension():
             [False, True, True],
             [True, True, True],
         ],
-    )
+    ).transpose(['yy', 'xx'])
 
     expected = da.assign_masks(m=mask).sum(["xx", "yy"])
     line = next(iter(fig1d.artists.values()))
@@ -298,25 +266,8 @@ def test_polygon_mode_triangle_with_mask_in_third_dimension():
 
 @pytest.mark.usefixtures('_use_ipympl')
 def test_polygon_mode_preserves_keep_dim_binedges():
-    da = sc.DataArray(
-        data=sc.array(
-            dims=["xx", "yy", "zz"],
-            values=[
-                [[0, 1, 2], [3, 4, 5], [6, 7, 8]],
-                [[9, 10, 11], [12, 13, 14], [15, 16, 17]],
-                [[18, 19, 20], [21, 22, 23], [24, 25, 26]],
-                [[27, 28, 29], [30, 31, 32], [33, 34, 35]],
-            ],
-            dtype=float,
-        ),
-        coords={
-            "xx": sc.array(
-                dims=['xx'], values=[0.0, 100.0, 200.0, 300.0, 400.0], unit='m'
-            ),
-            "yy": sc.array(dims=['yy'], values=[0.0, 10.0, 20.0, 30.0], unit='m'),
-            "zz": sc.array(dims=['zz'], values=[0.0, 1.0, 2.0, 3.0], unit='m'),
-        },
-    )
+    da = _make_test_data()
+    da.coords['zz'] = sc.array(dims=['zz'], values=[0.0, 1.0, 2.0, 3.0], unit='m')
 
     ip = pp.inspector(da, mode='polygon', dim='zz')
     fig2d = ip[0][0]
@@ -326,8 +277,8 @@ def test_polygon_mode_preserves_keep_dim_binedges():
 
     # This triangle should select the bottom left corner of the data.
     # Closing the polygon by repeating the first point.
-    x = [-1, 32, -1, -1]
-    y = [-1, -1, 350, -1]
+    x = [-1, -1, 350, -1]
+    y = [-1, 32, -1, -1]
     for xi, yi in zip(x, y, strict=True):
         tool.click(x=xi, y=yi)
 
@@ -339,7 +290,7 @@ def test_polygon_mode_preserves_keep_dim_binedges():
             [False, True, True],
             [True, True, True],
         ],
-    )
+    ).transpose(['yy', 'xx'])
 
     expected = da.assign_masks(m=mask).sum(["xx", "yy"])
     line = next(iter(fig1d.artists.values()))
@@ -349,25 +300,7 @@ def test_polygon_mode_preserves_keep_dim_binedges():
 
 @pytest.mark.usefixtures('_use_ipympl')
 def test_polygon_mode_contains_nothing():
-    da = sc.DataArray(
-        data=sc.array(
-            dims=["xx", "yy", "zz"],
-            values=[
-                [[0, 1, 2], [3, 4, 5], [6, 7, 8]],
-                [[9, 10, 11], [12, 13, 14], [15, 16, 17]],
-                [[18, 19, 20], [21, 22, 23], [24, 25, 26]],
-                [[27, 28, 29], [30, 31, 32], [33, 34, 35]],
-            ],
-            dtype=float,
-        ),
-        coords={
-            "xx": sc.array(
-                dims=['xx'], values=[0.0, 100.0, 200.0, 300.0, 400.0], unit='m'
-            ),
-            "yy": sc.array(dims=['yy'], values=[0.0, 10.0, 20.0, 30.0], unit='m'),
-            "zz": sc.array(dims=['zz'], values=[0.0, 1.0, 2.0], unit='m'),
-        },
-    )
+    da = _make_test_data()
 
     ip = pp.inspector(da, mode='polygon', dim='zz')
     fig2d = ip[0][0]
@@ -377,8 +310,8 @@ def test_polygon_mode_contains_nothing():
 
     # This triangle should select the bottom left corner of the data.
     # Closing the polygon by repeating the first point.
-    x = [-100, -1, -100, -100]
-    y = [-1, -1, 350, -1]
+    x = [-1, -1, 350, -1]
+    y = [-100, -1, -100, -100]
     for xi, yi in zip(x, y, strict=True):
         tool.click(x=xi, y=yi)
 
@@ -392,25 +325,7 @@ def test_polygon_mode_contains_nothing():
 
 @pytest.mark.usefixtures('_use_ipympl')
 def test_rectangle_mode():
-    da = sc.DataArray(
-        data=sc.array(
-            dims=["xx", "yy", "zz"],
-            values=[
-                [[0, 1, 2], [3, 4, 5], [6, 7, 8]],
-                [[9, 10, 11], [12, 13, 14], [15, 16, 17]],
-                [[18, 19, 20], [21, 22, 23], [24, 25, 26]],
-                [[27, 28, 29], [30, 31, 32], [33, 34, 35]],
-            ],
-            dtype=float,
-        ),
-        coords={
-            "xx": sc.array(
-                dims=['xx'], values=[0.0, 100.0, 200.0, 300.0, 400.0], unit='m'
-            ),
-            "yy": sc.array(dims=['yy'], values=[0.0, 10.0, 20.0, 30.0], unit='m'),
-            "zz": sc.array(dims=['zz'], values=[0.0, 1.0, 2.0], unit='m'),
-        },
-    )
+    da = _make_test_data()
 
     ip = pp.inspector(da, mode='rectangle', dim='zz')
     fig2d = ip[0][0]
@@ -419,9 +334,8 @@ def test_rectangle_mode():
     tool = fig2d.toolbar['inspect']._tool
 
     # This rectangle should select the bottom left corner of the data.
-    # Closing the rectangle by repeating the first point.
-    x = [-1, 19]
-    y = [-1, 290]
+    x = [-1, 290]
+    y = [-1, 19]
     for xi, yi in zip(x, y, strict=True):
         tool.click(x=xi, y=yi)
 
@@ -433,7 +347,7 @@ def test_rectangle_mode():
             [False, False, True],
             [True, True, True],
         ],
-    )
+    ).transpose(['yy', 'xx'])
 
     expected = da.assign_masks(m=mask).sum(["xx", "yy"])
     line = next(iter(fig1d.artists.values()))
@@ -442,25 +356,7 @@ def test_rectangle_mode():
 
 @pytest.mark.usefixtures('_use_ipympl')
 def test_rectangle_mode_contains_nothing():
-    da = sc.DataArray(
-        data=sc.array(
-            dims=["xx", "yy", "zz"],
-            values=[
-                [[0, 1, 2], [3, 4, 5], [6, 7, 8]],
-                [[9, 10, 11], [12, 13, 14], [15, 16, 17]],
-                [[18, 19, 20], [21, 22, 23], [24, 25, 26]],
-                [[27, 28, 29], [30, 31, 32], [33, 34, 35]],
-            ],
-            dtype=float,
-        ),
-        coords={
-            "xx": sc.array(
-                dims=['xx'], values=[0.0, 100.0, 200.0, 300.0, 400.0], unit='m'
-            ),
-            "yy": sc.array(dims=['yy'], values=[0.0, 10.0, 20.0, 30.0], unit='m'),
-            "zz": sc.array(dims=['zz'], values=[0.0, 1.0, 2.0], unit='m'),
-        },
-    )
+    da = _make_test_data()
 
     ip = pp.inspector(da, mode='rectangle', dim='zz')
     fig2d = ip[0][0]
@@ -468,10 +364,8 @@ def test_rectangle_mode_contains_nothing():
     fig2d.toolbar['inspect'].value = True
     tool = fig2d.toolbar['inspect']._tool
 
-    # This rectangle should select the bottom left corner of the data.
-    # Closing the rectangle by repeating the first point.
-    x = [-100, -1]
-    y = [-1, -1]
+    x = [-1, -1]
+    y = [-100, -1]
     for xi, yi in zip(x, y, strict=True):
         tool.click(x=xi, y=yi)
 
@@ -498,25 +392,7 @@ def test_creation_with_non_dimension_coord():
 )
 @pytest.mark.parametrize("mode", ["point", "polygon", "rectangle"])
 def test_different_operations(operation, mode):
-    da = sc.DataArray(
-        data=sc.array(
-            dims=["xx", "yy", "zz"],
-            values=[
-                [[0, 1, 2], [3, 4, 5], [6, 7, 8]],
-                [[9, 10, 11], [12, 13, 14], [15, 16, 17]],
-                [[18, 19, 20], [21, 22, 23], [24, 25, 26]],
-                [[27, 28, 29], [30, 31, 32], [33, 34, 35]],
-            ],
-            dtype=float,
-        ),
-        coords={
-            "xx": sc.array(
-                dims=['xx'], values=[0.0, 100.0, 200.0, 300.0, 400.0], unit='m'
-            ),
-            "yy": sc.array(dims=['yy'], values=[0.0, 10.0, 20.0, 30.0], unit='m'),
-            "zz": sc.array(dims=['zz'], values=[0.0, 1.0, 2.0], unit='m'),
-        },
-    ).transpose(['zz', 'yy', 'xx'])
+    da = _make_test_data()
 
     ip = pp.inspector(da, mode=mode, dim='zz', operation=operation)
     fig2d = ip[0][0]

--- a/tests/plotting/inspector_test.py
+++ b/tests/plotting/inspector_test.py
@@ -490,3 +490,51 @@ def test_creation_with_non_dimension_coord():
         x_alt=da.coords['x'] * 1.1, y_alt=da.coords['y'] * 1.1
     ).drop_coords(['x', 'y'])
     pp.inspector(da, coords=['x_alt', 'y_alt'])
+
+
+@pytest.mark.usefixtures('_use_ipympl')
+@pytest.mark.parametrize(
+    "operation", ["sum", "mean", "min", "max", "nansum", "nanmean", "nanmin", "nanmax"]
+)
+@pytest.mark.parametrize("mode", ["point", "polygon", "rectangle"])
+def test_different_operations(operation, mode):
+    da = sc.DataArray(
+        data=sc.array(
+            dims=["xx", "yy", "zz"],
+            values=[
+                [[0, 1, 2], [3, 4, 5], [6, 7, 8]],
+                [[9, 10, 11], [12, 13, 14], [15, 16, 17]],
+                [[18, 19, 20], [21, 22, 23], [24, 25, 26]],
+                [[27, 28, 29], [30, 31, 32], [33, 34, 35]],
+            ],
+            dtype=float,
+        ),
+        coords={
+            "xx": sc.array(
+                dims=['xx'], values=[0.0, 100.0, 200.0, 300.0, 400.0], unit='m'
+            ),
+            "yy": sc.array(dims=['yy'], values=[0.0, 10.0, 20.0, 30.0], unit='m'),
+            "zz": sc.array(dims=['zz'], values=[0.0, 1.0, 2.0], unit='m'),
+        },
+    ).transpose(['zz', 'yy', 'xx'])
+
+    ip = pp.inspector(da, mode=mode, dim='zz', operation=operation)
+    fig2d = ip[0][0]
+    fig1d = ip[0][1]
+    fig2d.toolbar['inspect'].value = True
+    tool = fig2d.toolbar['inspect']._tool
+
+    match mode:
+        case 'point':
+            x = [105.0]
+            y = [5.0]
+        case 'rectangle':
+            x = [52.0, 189.0]
+            y = [17.0, 24.0]
+        case 'polygon':
+            x = [105.0, 256.0, 256.0, 105.0, 105.0]
+            y = [5.0, 5.0, 15.0, 15.0, 5.0]
+    for xi, yi in zip(x, y, strict=True):
+        tool.click(x=xi, y=yi)
+
+    assert len(fig1d.artists) == 1

--- a/tests/plotting/slicer_test.py
+++ b/tests/plotting/slicer_test.py
@@ -231,6 +231,15 @@ class TestSlicer1d:
         da = da.assign_coords(x_alt=da.coords['xx'] * 2).drop_coords('xx')
         SlicerPlot(da, keep=['x_alt'], mode='single', coords=['x_alt'])
 
+    @pytest.mark.parametrize(
+        "operation",
+        ["sum", "mean", "min", "max", "nansum", "nanmean", "nanmin", "nanmax"],
+    )
+    @pytest.mark.parametrize("mode", ["single", "range", "combined"])
+    def test_different_operations(self, operation, mode):
+        da = data_array(ndim=2)
+        SlicerPlot(da, keep=['xx'], mode=mode, operation=operation)
+
 
 @pytest.mark.usefixtures("_parametrize_interactive_2d_backends")
 class TestSlicer2d:
@@ -389,3 +398,12 @@ class TestSlicer2d:
         SlicerPlot(
             da, keep=['y_alt', 'x_alt'], mode='single', coords=['x_alt', 'y_alt']
         )
+
+    @pytest.mark.parametrize(
+        "operation",
+        ["sum", "mean", "min", "max", "nansum", "nanmean", "nanmin", "nanmax"],
+    )
+    @pytest.mark.parametrize("mode", ["single", "range", "combined"])
+    def test_different_operations(self, operation, mode):
+        da = data_array(ndim=3)
+        SlicerPlot(da, keep=['xx', 'yy'], mode=mode, operation=operation)

